src/transform/src/column_knowledge.rs

```diff
@@ -111,49 +111,71 @@ impl ColumnKnowledge {
                     Ok(body_knowledge)
                 }
                 MirRelationExpr::LetRec { ids, values, body } => {
-                    // As a first approximation, we treat LetRec blocks as
-                    // barriers for this optimization.
+                    let mut max_iterations = 0;
 
-                    // A map to hold knowledge from shadowed bindings
-                    let mut shadowed_knowledge = BTreeMap::new();
-
-                    // We set knowledge[i][j] = DatumKnowledge::empty() for each
-                    // column j and CTE i.
+                    // Set knowledge[i][j] = DatumKnowledge::bottom() for each
+                    // column j and CTE i. This corresponds to the normal
+                    // evaluation semantics where each recursive CTE is
+                    // initialized to the empty collection.
                     for (id, value) in zip_eq(ids.iter(), values.iter()) {
                         let id = mz_expr::Id::Local(id.clone());
-                        let knowledge_new = vec![DatumKnowledge::empty(); value.arity()];
-                        if let Some(knowledge_old) = knowledge.insert(id, knowledge_new) {
-                            shadowed_knowledge.insert(id, knowledge_old);
-                        };
+                        let knowledge_new = vec![DatumKnowledge::bottom(); value.arity()];
+                        max_iterations += knowledge_new.len();
+                        let prior_knowedge = knowledge.insert(id, knowledge_new);
+                        assert!(prior_knowedge.is_none());
                     }
 
-                    // For the first loop, we sequentially absorb into
-                    // knowledge[i][j] the result of descending into values[i].
+                    // TODO: This should respect the soft and hard max iteration
+                    // counts that will be enfoced when actually evaluating the
+                    // loop (see #18362 and #16800).
+                    max_iterations *= std::cmp::min(4 * max_iterations, 100);
```

Comment:
Did you mean to write `=` instead of `*=`?

Why multiply with 4? I would think the knowledge about each of the columns has to reach `Any(true)` in at most `3` steps, due to the height of the lattice.

I think if we don't reach a fixpoint in `3 * total_arity`, then something is very wrong with our understanding of the situation. So maybe log a sentry warning when that happens? (But then remove the `min` with the `100`.) What do you think?

Reply:
Miscounted by one because we have a longest path of states and 3 jumps between them, but we actually initialize with `bottom`, so you're right that we need at most 3 steps to reach the top.

Reply:
> But then remove the min with the 100.

Summonning @antiguru because this is more a PM decision.

If I remove the hard max here
```rust
max_iterations *= std::cmp::min(3 * max_iterations, 100);
```
we are guaranteed to converge and be optimal in all cases, but in some pathological cases with a lot of wide columns the optimization time might be heavily penalized (especially at the moment since we need to clone every recursive `value` binding at every iteration).

If I keep it unchanged we move forward with the most conservative approximation (essentially giving up on inferring any knowledge).

Reply:
I think the loop exit conditions are now as agreed above.